doc/SAI-Counter-enhancement-hld.md

```diff
@@ -0,0 +1,470 @@
+# SAI Counter enhancement
+-------------------------------------------------------------------------------
+ Title       | SAI counter enahncement
+-------------|-----------------------------------------------------------------
+ Authors     |<code> Rajkumar P R, Marvell Technology Inc <br>           Ravindranath C K, Marvell Technology Inc </code>
+ Status      | In review
+ Type        | Standards track
+ Created     | 01/07/2024
+
+-------------------------------------------------------------------------------
+
+## Problem Statement
+Within the current framework of SAI, objects such as vlan, router-interface, tunnel, queue, port, and others come with built-in statistical enumeration. The expectation is that these statistics will be counted by default upon the creation of objects. This proposal contributes to achieving efficient utilization of flexible counters based on the counting requirements of deployments. It enables platforms operating within scaled environments to seamlessly map these flexible counters, thereby facilitating the realization of this objective.
+
+## Overview
+When some SAI objects are created, the default statistics enum defined in each these object file are expected to be counted. Current model has no option to disable/enable statistics counting. As part of this proposal, we would like to add the flexibility to manage the counting at run-time. With this,
+- User has flexibility to decide which object should support counting.
+- Helps to optimally use the hardware counting resources.
+
+## Options
+1. A new SAI attribute introduced in each of these object to determine counting mode.
+2. In parallel to existing design, introduce a new counter ID attribute per statistics type to count packet, byte.
+
+## SAI Spec Enhancement
+### Option 1
+
+**New enum defined for the counting mode.**
+```c
+typedef enum _sai_stats_count_mode_t
+{
+ 
+    /** Count packet and byte */
+    SAI_STATS_COUNT_MODE_PACKET_AND_BYTE,
+
+    /** Count only packet */
+    SAI_STATS_COUNT_MODE_PACKET,
+
+    /** Count only byte */
+    SAI_STATS_COUNT_MODE_BYTE,    
+
+     /** Counting is disabled */
+    SAI_STATS_COUNT_MODE_NONE
+
+} sai_stats_count_mode_t;
```

Comment:
Should we leverage sai_counter_stat_t and add these new enums to sai_counter_stat_t ? This enum is similar to sai_counter_stat_t.

Reply:
We should not mix count mode with sai_counter_stat_t.